python/ray/util/spark/cluster_init.py

```diff
@@ -0,0 +1,684 @@
+            try:
+                get_dbutils().entry_point.registerBackgroundSparkJobGroup(
+                    spark_job_group_id
+                )
+            except Exception:
+                _logger.warning(
+                    "Register ray cluster spark job as background job failed. You need to manually "
+                    "call `ray_cluster_on_spark.shutdown()` before detaching your databricks "
+                    "python REPL."
+                )
+        return ray_cluster_handler
+    except Exception:
+        # If driver side setup ray-cluster routine raises exception, it might result in part of ray
+        # processes has been launched (e.g. ray head or some ray workers have been launched),
+        # calling `ray_cluster_handler.shutdown()` to kill them and clean status.
+        ray_cluster_handler.shutdown()
+        raise
+
+
+_active_ray_cluster = None
+
+
+@PublicAPI(stability="alpha")
+def init_ray_cluster(
+    num_worker_nodes=None,
+    total_cpus=None,
+    total_gpus=None,
+    total_heap_memory_bytes=None,
+    total_object_store_memory_bytes=None,
+    heap_to_object_store_memory_ratio=None,
+    head_options=None,
+    worker_options=None,
+    ray_temp_root_dir="/tmp",
+    ray_log_root_dir="/tmp",
+    safe_mode=True,
+):
+    """
+    Initialize a ray cluster on the spark cluster by starting a ray head node in the spark
+    application's driver side node.
+    After creating the head node, a background spark job is created that
+    generates an instance of `RayClusterOnSpark` that contains configuration for the ray cluster
+    that will run on the Spark cluster's worker nodes.
+    After a ray cluster initialized, your python process automatically connect to the ray cluster,
+    you can call `ray.util.spark.shutdown_ray_cluster` to shut down the ray cluster.
+    Note: If the active ray cluster haven't shut down, you cannot create a new ray cluster.
+
+    Args
+        num_worker_nodes: The number of spark worker nodes that the spark job will be submitted to.
+            This argument represents how many concurrent spark tasks will be available in the
+            creation of the ray cluster. The ray cluster's total available resources
+            (memory, CPU and/or GPU)
+            is equal to the quantity of resources allocated within these spark tasks.
+            Specifying the `num_worker_nodes` as `-1` represents a ray cluster configuration that
+            will use all available spark tasks slots (and resources allocated to the spark
+            application) on the spark cluster.
+            To create a spark cluster that is intended to be used exclusively as a shared ray
+            cluster, it is recommended to set this argument to `-1`.
+        total_cpus: The total cpu core count for the ray cluster to utilize.
+        total_gpus: The total gpu count for the ray cluster to utilize.
+        total_heap_memory_bytes: The total amount of heap memory (in bytes) for the ray cluster
```

Comment:
Because we make one spark task map to one ray worker node,
The `cpus_per_ray_worker_node` / `gpus_per_ray_worker_node` are fixed by spark application config. i.e. `cpus_per_ray_worker_node` must equal to spark application config `spark.task.cpus` and `gpus_per_ray_worker_node` must equal to spark application config `spark.task.resource.gpu.amount`. So setting them is meaningless.

Reply:
To simplify interface, we can only provide `num_worker_nodes` parameter, and it results in launching a spark job with task number equals to `num_worker_nodes`, and resources (cpu / gpu / memory) allocated to each spark task has been fixed by spark application config, so it already can determine the whole ray cluster total resources.

Reply:
I think only allow `num_worker_nodes` is probably a good simplification, since the user has to be doing some of this calculation anyways. We can document that you should use the Spark configs to control the resource allocation per worker node.

Reply:
CC @jjyao to confirm https://github.com/ray-project/ray/pull/28771#discussion_r1035180475

Reply:
if we only set `num_worker_nodes` does it mean the ray cluster could always use the full resource configured by `spark.task.cpus` and `spark.task.resource.gpu.amount` ?

Reply:
> if we only set num_worker_nodes does it mean the ray cluster could always use the full resource configured by spark.task.cpus and spark.task.resource.gpu.amount ?

Yes. In this case, it assumes each ray worker node is configured with `spark.task.cpus` cores, and `spark.task.resource.gpu.amount` GPUs.